standards/core/rln-contract.md

```diff
@@ -0,0 +1,301 @@
+Memberships MUST NOT be transferable.
+A user MAY use one Ethereum address to manage multiple memberships.
+A user MAY use one Waku node to manage multiple memberships. [^1]
+
+[^1]: No Waku implementation supports managing multiple memberships from one node (as of August 2024).
+
+
+## Contract functionalities
+
+Availability of membership-specific functionalities MUST be as follows:
+
+|                       | Active | GracePeriod | Expired | ErasedAwaitsWithdrawal | Erased |
+| --------------------- | ------ | ----------- | ------- | ---------------------- | ------ |
+| Send a message        | Yes    | Yes         | Yes     | No                     | No     |
+| Extend the membership | No     | Yes         | No      | No                     | No     |
+| Withdraw the deposit  | No     | Yes         | Yes     | Yes                    | No     |
+
+Sending a message is included here for completeness,
+although it is part of the RLN Relay protocol and not the RLN contract.
+
+### Register a membership
+
+Membership registration is subject to the following conditions:
+- if there are _Expired_ memberships in the RLN tree, the new membership MUST overwrite an _Expired_ membership;
+- the new membership SHOULD overwrite the membership that had been _Expired_ for the longest time;
+- if a new membership A overwrites an _Expired_ membership B:
+	- membership B MUST become _ErasedAwaitsWithdrawal_;
+	- the current total rate limit MUST be decremented by the rate limit of membership B;
+	- the contract MUST take all necessary steps to ensure that the holder of membership B can withdraw their deposit later;
+- registration MUST fail if the total rate limit of _Active_, _GracePeriod_, and _Expired_ memberships, including the one being created, would exceed the maximum total rate;
+- registration MUST fail if the requested rate limit for a new membership is lower than the minimal allowed rate limit;
+- the user MUST lock-up a deposit to register a membership;
+- the user MUST specify the rate limit of the new membership[^2];
+- the size of the deposit MUST depend on the requested rate limit;
+- in case of a successful registration:
+	- the new membership MUST become _Active_;
+	- the current total rate limit MUST be incremented by the rate limit of the new membership;
+- a newly created membership MUST have an expiration time `T` and a grace period `G`.
+
+[^2]: A user-facing application SHOULD suggest default values for rate limits for the user.
+
+### Extend a membership
+
+Extending a membership is subject to the following conditions:
+- extension MUST fail if the membership is in any state other than _GracePeriod_;
+- the membership holder MUST be able to extend their membership;
+- any user except the membership holder MUST NOT be able to extend a membership;
+- after a successful extension, the membership MUST become _Active_.
+
+Holding a membership means controlling the private key from which the RLN commitment ID (i.e., public key) was derived.
```

Comment:
_private key/public key/derive_ are not term we usually used for RLN, I am worried it can bring confusion.
It's usually _secret/commitment/(poseidon) hash_

> the membership holder MUST be able to extend their membership;

You are saying that the smart contract should determine membership holder based on the RLN commitment. Meaning that some proof of knowing the secret to the RLN commitment should be passed to extend.

I'd suggest to simplify this and assume that from a smart contract pov, the membership holder is the smart contract/ethereum account that send the registration transaction.

Reply:
Also move this definition under "Register a membership" and have a bullet point:

> - the membership holder is defined as....

Reply:
Yes it absolutely make sense and indeed, the choice is not as trivial as I first thought.

It seems that the choice is what the user is most likely to **not loose**. Ethereum wallet seed or RLN secret/keystore.

When  withdrawing a deposit, does the deposit go back to original address or to address used to initiate the withdraw transaction? I assume latter.

I think it is fair to assume that the user secured their Ethereum wallet and do not loose access to it.

I do not expect this to be as clear cut for the RLN keystore.

If a user install a Waku app, gets the RLN membership, then desinstall/delete/change device. keystore maybe lost.

Whereas the wallet they used may still be accessible, or even if they use status wallet, they are expected to backup their seed and potentially import it in another wallet. They can then get their deposit back using a web portal.
Whereas they are unlikely to be able to import their RLN keystore in another app.

At this point in time, I would say it would be more fool proof to assume the security of the wallet, not the RLN keystore, and hence assume ownership based on the ethereum address used for registration.

We can then evolve that, and possibly allow either. We will also have to review subsidization scenarios.
If project X subsidize RLN membership for users, then user should not be able to withdraw deposit.
Payer and user may be different entities. Which is an argument to enable deposit withdraw on ethereum address.

WDYT @jm-clius ?

Reply:
I'd also add that the answer may depend on whether we're thinking more in terms of a service provider model, or a model where end-users manage their own memberships. In the former case, I'd expect professional entities to be more likely to e.g. rotate Ethereum addresses, have multiple addresses for cold- and hot-storage, etc. That for me indicates that a separate RLN-level identity is more suitable. In the latter case, I'd agree that the end user is more likely to lose their RLN seed phrase than their Ethereum seed phrase, as their Ethereum address might also hold value unrelated to RLN.

Reply:
>  I'd expect professional entities to be more likely to e.g. rotate Ethereum addresses, have multiple addresses for cold- and hot-storage, etc

Keen to get infra's thought on this.

To summarize for infra, in terms of withdrawing funds from the RLN smart contract, either:

A. The Ethereum account originally used to register a membership must be used to withdraw the funds.

OR

B. Any Ethereum address can be used, but a zk-proof to own the RLN membership must be passed (proof to be generated with a provided script etc).


Side note, B is likely to make gas cost higher as more data is needed to be passed.

Reply:
I always prefer simpler solutions, both to use and understand, so I'd go with A.

Reply:
OK, then I'm introducing these definitions:
- a _holder_ is the one who controls the RLN secret;
- a _keeper_ is the one who controls the Ethereum address;
- the contract only uses keeper / non-keeper as authorization criterion (that is, we only allow extending and withdrawing from the same Ethereum address; no RLN proof necessary).

Reply:
To make the definition generic for AA, EOA etc, I believe the right terminlogy would be: the _keeper_ is the _sender_ of the transaction registering the membership.

Reply:
Fair pint. I'm not sure however that the term "sender" is clear enough, as it might mean:

- `msg.sender` in Solidity semantics;
- the one who pays for gas (paymaster);
- the one who first broadcasts the transaction into the network (which is Infura in most cases).

What we need here is a cryptographic identity that would link a user to their membership from registration to withdrawal. The cleanest way to do so would be to use our app-specific identity (aka RLN ID), but as a temporary solution to simplify initial implementation, we're piggy-backing on the Ethereum addresses as identity provider. I'm not sure we can guarantee that this scheme will work for all future Ethereum developments, in particular the de-coupling of various services such as transaction signing, gas payment, broadcasting, etc.

Given that, would it make sense to explicitly state that we only support EOAs for the inintial implementation?

Reply:
By _sender_ I mean exactly `msg.sender` in solidity semantics.

Reply:
>  I'm not sure we can guarantee that this scheme will work for all future Ethereum developments

I would expect we are not the only one with this problem and it may be worth investigating, at some point, what is done in the ecosystem. AFAIK, the usage of `msg.sender` is the regular practice. This, or usage of signature such as EIP-712.

Reply:
Added a [clarification](https://github.com/waku-org/specs/pull/34/commits/818ce9847672c6b27a91c81d0c6794224179cf24).